include/seastar/http/request.hh

```diff
@@ -109,15 +111,90 @@ struct request {
 
     /**
      * Search for the last query parameter of a given key
-     * @param key the query paramerter key
+     * @param key the query parameter key
      * @return the query parameter value, if it exists or empty string
      */
     sstring get_query_param(const sstring& key) const {
-        auto res = query_parameters.find(key);
-        if (res == query_parameters.end()) {
-            return "";
+        if(auto res = query_parameters.find(key); res != query_parameters.end()) {
+            return res->second;
         }
-        return res->second;
+
+        auto res = query_params.find(key);
+        if (res != query_params.end()) {
+            return res->second.empty() ? "" : res->second.back();
+        }
+
+        return "";
+    }
+
+    /**
+     * Search for all query parameters of a given key
+     * @param key the query parameter key
+     * @return a vector of all query parameter values, if it exists or an nullopt
+     */
+    std::optional<std::vector<sstring>> get_query_param_array(const sstring& key) const {
+        std::vector<sstring> res;
+        if (auto it = query_params.find(key); it != query_params.end()) {
+            res = it->second;
+        }
+
+        if (auto old_it = query_parameters.find(key); old_it != query_parameters.end()) {
+            res.push_back(old_it->second);
+        }
+
+        if (res.empty()) {
+            return std::nullopt;
+        }
+
+        return res;
+    }
+
+    /**
+     * Set a query parameter value
+     * @param key the query parameter key
+     * @param value the query parameter value
+     */
+    void set_query_param(const sstring& key, const sstring& value) {
+        query_parameters[key] = value;
+        if (auto it = query_params.find(key); it != query_params.end()) {
+            query_params.erase(it);
+        }
+    }
+
+    /**
+     * Set query parameters in the request objects.
+     * @param params a map of query parameters, where the key is the parameter name and the value is a vector of values.
+     */
+    void set_query_params(const std::unordered_map<sstring, std::vector<sstring>>& params) {
```

Comment:
I don't quite like this method. You walk the params anyway, so it can be obsoleted by the

```c++
    request& set_query_params(std::string_view key, std::initializer_list<std::string_view>> values) {
        ...; // assign query_params[key] = values and query_parameters[key] = tail(values)
        return *this;
    }
    request& set_query_params(std::string_view key, std::string_view val) {
        return set_query_params(key, {val});
    }
```

and then use it like

```c++
    req.set_query_params("foo", "bar").set_query_params("foo", {"buz", "fux"});
```